### PR TITLE
Add metric to track Unhealthy Stale machine removal count

### PR DESCRIPTION
**Pull request by jguipi**
**What this PR does / why we need it**:
There are two kinds of stale machine which `machineSet controller` collects 
- stale due to unhealthiness
- stale due to creation timeout

In this metric , we are interested to expose the count of `stale due to unhealthiness` machine which get terminated between two scrapings by prometheus.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

I have tested the code for different scenarios like:
- deleting a machine obj directly doesn't increase count
- health timeout machines in different machineDeployments don't cause race and update the counter properly
- restarting mcm resets counter back to 0
- stale machines due to creation timeout collection doesn't increase the counter

**Release note**:
```other operator
Added a new metric that will allow to get the number of stale (due to unhealthiness) machines  that are getting terminated
```

### CODE REVIEW

**Comment by jguipi** [comment]
/area monitoring
/kind enhancement

Here's a capture of the exported metric visibility : 

image

cc.: @dmahmalatsap @istvanballok @rickardsjp @himanshu-kun @rishabh-11

**Review by himanshu-kun — CHANGES_REQUESTED** [review]
there are many worker go-routines which reconcile the machineSets , and the current metric is per machineSet
So I don't feel this is a thread safe way of keeping count , I'll have to discuss internally and research and only then I can comment on that.
otherwise I have requested some changes.

**Comment by elankath** [comment]
nice work.

File: pkg/controller/machineset.go [hunk 1]
```diff
@@ -345,6 +357,7 @@ func (c *controller) manageReplicas(ctx context.Context, allMachines []*v1alpha1
 
 	if len(staleMachines) >= 1 {
 		klog.V(2).Infof("Deleting stale machines")
+		staleMachineCounter += len(staleMachines)
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/controller/machineset.go]: I have addressed these in the new commits I added

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -59,6 +59,15 @@ var (
 		Help:      "Information of the mcm managed Machines' status conditions.",
 	}, []string{"name", "namespace", "condition"})
 
+	// StaleMachineCount Number of stale (failed) machines that get flagged for termination
+	StaleMachineCount = prometheus.NewCounterVec(prometheus.CounterOpts{
```
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/metrics/metrics.go]: The current code design doesn't allow to add this metric per machineSet / machineDeployment easily, as we would need separate variables to track count for each machineSet/machineDeployment along with thread safe read/write. 
@jguipi hope a metric with cluster-wide scope is fine for now.

**Comment by dmahmalatsap** [comment]
Thanks for the refactor and for keeping us in mind in the process !

File: pkg/metrics/metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (himanshu-kun) [reply pkg/metrics/metrics.go]: sure sounds good

**Review by rishabh-11 — APPROVED** [review]
Looks good. thanks for the changes.

/lgtm

**Review by himanshu-kun — APPROVED** [review]
/lgtm
approving my own contribution to the PR :')

**Comment by himanshu-kun** [comment]
Note: Since this metric is expose by MCM core, so it doesn't require to be vendored in mcm-providers to start exposing the metrics and will be available in next MCM release (v0.50.0)

**Pull request merged**